namerd/iface/control-http/src/main/scala/io/buoyant/namerd/iface/HttpControlService.scala

```diff
@@ -225,4 +257,101 @@ class HttpControlService(storage: DtabStore) extends Service[Request, Response]
       case Throw(_: DtabNamespaceDoesNotExistException) => Future.value(Response(Status.NotFound))
       case Throw(_) => Future.value(Response(Status.InternalServerError))
     }
+
+  private[this] def streamingResp[T](values: Event[T], render: (T, Closable) => Buf): Future[Response] = {
+    val resp = Response()
+    resp.setChunked(true)
+    val writer = resp.writer
+    var closable: Closable = null
+    var writeFuture: Future[Unit] = Future.Unit
+    closable = values.respond { t =>
+      writeFuture = writeFuture.before {
+        writer.write(render(t, closable)).onFailure { _ =>
+          closable.close()
+        }
+      }
+    }
+    Future.value(resp)
+  }
+
+  private[this] val bindingCacheMu = new {}
+  private[this] var bindingCache: Map[(String, Path), Activity[NameTree[Name.Bound]]] = Map.empty
+  private[this] def getBind(ns: String, path: Path): Activity[NameTree[Name.Bound]] =
+    bindingCacheMu.synchronized {
+      val key = (ns, path)
+      bindingCache.get(key) match {
+        case Some(act) => act
+        case None =>
+          val act = namers(ns).bind(Dtab.empty, path)
+          bindingCache += (key -> act)
+          act
+      }
+    }
```

Comment:
I think this (and similar code elsewhere) could be simplified using a `java.util.concurrent.ConcurrentHashMap` (or alternately one of the Guava cache implementations, but I think `ConcurrentHashMap` gives us everything we want). The Java8 version of CHM includes a `computeIfAbsent` method that implements exactly the logic we have here, without the need for any extra `synchronized` blocks.